14765 add auth design document (#15008)
CDCgov/prime-reportstream: prime-router/docs/design/design/auth/auth-design.md

# Authentication and Authorization design

## Current design

We currently use a variety of methods to authenticate users and authorize actions in our system. 
We use a mix/match of the following for different endpoints:

- Okta
  - A cloud-hosted identity and access management service
- Server2Server
  - Homegrown JWT-based authn/authz
- Azure Function Keys
  - Azure configured shared secrets for authorization
  - Does not handle authentication at all! (Any user can use any key and the system will work)

## Reasons for updating

- Unify authn/authz to one consistent method as having multiple methods makes it difficult to determine who is using which method
- Reduce attack surface area since currently we have the weaknesses of each method
- Minimize authn/authz issues being the cause for missing sent reports
- Make life easier for Engagement team by having a single choice when onboarding a new sender/receiver

## Requirements

> - Protect certain endpoints from only being accessed by those we trust

Reviewer: nit

```suggestion
- Protect certain endpoints from only being accessed by identified trusted entities
```

- Centralize user management
- Revoke access if a secret is compromised
- Audit which clients are accessing which resources
- Fine-tune access control for specific clients
- Unify authn/authz approach for all microservices
- Allow clients the ability to self-service to ease client onboarding

## Identity Provider Options

### Azure AD

Pros
- Integrated into the Microsoft ecosystem that we currently reside
- Scales well

Cons
- Complex setup
 
### Okta

Pros
- A trusted industry standard
- Highly scalable
- Offers self-service to ease Engagement teams workload
- We already have many of our clients set up here!

Cons
- Single point of failure if there is an outage

### Homegrown

Pros
- As flexible as we need it to be
- Cheap

Cons
- Custom authn/authz code is often vulnerable to attacks we may not have considered
- Additional work for us to write and maintain

## Protocol

> We will be using OAuth 2.0 with OIDC. OAuth 2.0 is an industry-standard authorization protocol and OIDC is

Reviewer: I'm not sure OIDC is applicable here since it is for authenticating USERS and backend will deal with machine-machine auth that is not tied to a particular user. OIDC is used presently by the front end for user auth for website, however, if that's what you meant.

Author: I believe it is still technically OIDC but perhaps I should expand on this in the doc.

My thought is using the [client credentials flow](https://developer.okta.com/docs/concepts/oauth-openid/#client-credentials-flow) which seems like the right approach.

another open standard on top of OAuth that adds an identity layer (authentication).

Azure and Okta both support this protocol but if we were to go with the homegrown I would still highly recommend 
following the spec.

## Architecture options

### Spring reverse proxy microservice

This approach would create a new microservice whose sole responsibility is authn/authz. It would sit in front of
all requests to protected endpoints.

### Prerequisites
- Designate protected endpoints and put them behind a VPC.
  - Traffic must flow through our authn/authz service  
  - Protected endpoints should not be accessible to the public internet

A typical request flow would look like this
- Incoming request to protected endpoint hits our authn/authz service
- Depending on our implementation, we decide if that request is legitimate and allowed to access the designated endpoint
  - In cloud-based solutions, we would make calls out to an external IdP
  - In a homegrown solution, we would introspect certain cookies and headers and compare them to what we have in a datastore

> - If the request is unauthenticated, forward the user along to a login page (often provided by the external IdP)

Reviewer: This will not work for machine-machine auth.

Author: Yup let me update this for the right flow.

- If the request is unauthorized, short-circuit the request to a 401 unauthorized
- If the request is authenticated and authorized for the resource, pass it along to the internal protected endpoint

> Pros

Reviewer: Can this reverse proxy be configured for other endpoints outside of core RS microservices, like ETOR endpoints? If so, let's add that as a pro because it would allow RS users and machines to access ETOR endpoints without having to auth separately

Author: It can be configured for anything we want!

This approach would attempt to use existing libraries to annotate request mappings in our Spring app.

We would have to write custom code in the shared project to handle configuration and authn/authz for Report Stream
endpoints that still live in our Azure Functions app.

Pros
- Simple implementation for a small project like ours
- Authn/authz annotations would be directly adjacent to the endpoint mappings they protect
- Can handle complex authorization requirements for specific endpoints with more ease
- Simplifies project structure without new deployments
- No additional devops support

Cons
- Less scalable solution
- Would have to deal with handling authn/authz with 2 different web frameworks
  -  Easy for Spring but could require some custom work for Azure Functions

Code example:
```kotlin

// has read scope
@GetMapping("/api/v1/protected/endpoint")
@PreAuthorize("hasAuthority('SCOPE_read')")  
fun protectedReadEndpoint() {
    // read that anyone with the read scope can see
}

// has admin authority
@PostMapping("/api/v1/protected/endpoint")
@PreAuthorize("hasAuthority('Admin')")
fun protectedWriteEndpoint() {
    // write that only admins can do
}
```

> ## Definitions

Reviewer: Move this up right before architecture options section?

# Authentication and Authorization design

## Current design

We currently use a variety of methods to authenticate users and authorize actions in our system. 
We use a mix/match of the following for different endpoints:

- Okta
  - A cloud-hosted identity and access management service
- Server2Server

>   - Homegrown implementation of client credentials OIDC flow

Author: update to report stream implementation rather than homegrown

- Azure Function Keys
  - Azure configured shared secrets for authorization
  - Does not handle authentication at all! (Any user can use any key and the system will work)

## Reasons for updating

- Unify authn/authz to one consistent method as having multiple methods makes it difficult to determine who is using which method
- Reduce attack surface area since currently we have the weaknesses of each method
- Minimize authn/authz issues being the cause for missing sent reports
- Make life easier for Engagement team by having a single choice when onboarding a new sender/receiver

## Requirements

> - Protect certain endpoints from only being accessed by identified trusted entities

Author: Break down to more specific groups and endpoints

- Centralize user management
- Revoke access if a secret is compromised
- Audit which clients are accessing which resources
- Fine-tune access control for specific clients
- Unify authn/authz approach for all microservices

> - Allow clients the ability to self-service to ease client onboarding

Author: define this workflow

## Options discussed in 14765

- Azure Application Gateway
  - This is Azure's load balancer and does not handle authn/authz
- Azure Front Door
  - This is Azure's CDN technology and does not handle authn/authz
- Azure Identity Management
  - This is Azure's own implementation of handling authn/authz for Microsoft products and is not available for implementation in our application

## Identity Provider Options

### Azure AD

Pros
- Integrated into the Microsoft ecosystem that we currently reside
- Scales well

> Cons

Author: requires coordination with the CDC

- Complex setup
 
### Okta

Pros
- A trusted industry standard
- Highly scalable
- Offers self-service to ease Engagement teams workload
- We already have many of our clients set up here!

Cons
- Single point of failure if there is an outage

### Homegrown

Pros
- As flexible as we need it to be
- Cheap

Cons
- Custom authn/authz code is often vulnerable to attacks we may not have considered
- Additional work for us to write and maintain

## Protocol

We will be using OAuth 2.0 with OIDC. OAuth 2.0 is an industry-standard authorization protocol and OIDC is
another open standard on top of OAuth that adds an identity layer (authentication).

> We will be using the **client credentials** flow which was designed with machine-to-machine communication in 

Author: Add additional diagram for user flows.

mind. See the image below for how the flow works generally.

![Client Credentials Flow](clientCredentialsFlow.png)

## Definitions

- **authn**: Authenticate/authentication
- **authz**: Authorize/authorization
- **IAM**: Identity and access management
- **IdP**: Identity Provider
- **Okta**: A cloud-hosted identity and access management service

> - **Server2Server**: Homegrown JWT authn/authz

Reviewer 2: `ReportStream's implementation of JWT`...

A typical report request would look like this
- Client makes request to authorization server (or us in the homegrown solution) to retrieve token with their own credentials
- Client uses retrieved token with request to Report Stream
- Request comes in and hits our auth microservice instead of Report Stream directly
- Depending on our implementation, we decide if that request is legitimate and allowed to access the designated endpoint
  - In cloud-based solutions, we would make calls out to an external authz server
    - We could also introspect the token locally if speed is a factor but its more code to write/maintain
  - In a homegrown solution, we would introspect certain cookies and headers and compare them to what we have in a datastore
- If the request is unauthenticated or unauthorized, short-circuit the request to a 401
- If the request is authenticated and authorized for the resource, pass it along to the internal protected endpoint

Pros
- Extremely scalable
- All authn/authz code is in one place
- Security concerns can be centralized as all requests go through a single public endpoint
- Simplifies business logic as it no longer has to worry about authn/authz

> - Extensible outside of core RS services if required

Author: Check licensing for multiple applications and cost per user.

Cons
- Highly complex setup from both an engineering and devops perspective
- New service needs to be written and maintained
- A single point of failure if the service goes down
- Can lack customization options for very specific use-cases

Code example:
```kotlin

// handles all incoming requests
@RequestMapping("/**")
fun protectedReadEndpoint(
    @RequestBody(required = false) body: String,
    method: HttpMethod, 
    request: HttpServletRequest,
    response: HttpServletResponse
) {
    // this is all psudeocode!
    
    // check if we need to auth at all
    val authRequired = condigService.isAuthRequired(request.path, method)
    
    return if (authRequired) {
        // call out to auth service
        val authResult = authService.checkToken(request)

>         

Author: add authorization step

Make a example config file with both authn/authz

- If the request is unauthenticated or unauthorized, short-circuit the request to a 401
- If the request is authenticated and authorized for the resource, pass it along to the internal protected endpoint

Pros
- Extremely scalable
- All authn/authz code is in one place
- Security concerns can be centralized as all requests go through a single public endpoint
- Simplifies business logic as it no longer has to worry about authn/authz
- Extensible outside of core RS services if required

Cons
- Highly complex setup from both an engineering and devops perspective
- New service needs to be written and maintained
- A single point of failure if the service goes down
- Can lack customization options for very specific use-cases

Code example:
```kotlin

// handles all incoming requests
@RequestMapping("/**")
fun protectedReadEndpoint(
    @RequestBody(required = false) body: String,
    method: HttpMethod, 
    request: HttpServletRequest,
    response: HttpServletResponse
) {
    // this is all psudeocode!
    
    // check if we need to auth at all

>     val authRequired = condigService.isAuthRequired(request.path, method)

Author: typo

Reviewer 3: Assuming "condigService" was the typo, were you going to fix it?

Reviewer: occurs in 3 places FYI

    
    return if (authRequired) {
        // call out to authn/authz service
        val authnResult = authService.checkToken(request)
        val authzResult = authzService.checkAuthorization()
        
        // possible results
        when {
            authnResult == Authenticated && authzResult == Authorized -> httpClient.request(method, protectedEndpoint, body)
            else -> UnauthorizedResponse
        }
    } else {
        // pass through if no auth needed
        httpClient.request(method, protectedEndpoint, body)
    }
}
```

Authz configuration file example:

> ```json

Author: Config options to match client id

Author: claim vs scope

Author: add potential option of auth service not as a proxy

Author: investigate if its just overkill to match up client id given our users are trusted

Author: Add the options for how senders request their access token

Author: options on visibility outside organization

- Client uses retrieved token with request to ResportStream
- Request hits authentication microservice
  - Microservice calls out to Okta (or other provider) to check the validity of the access token and passes request along to ReportStream
  - If the request is unauthenticated, short-circuit the request to a 401
- Request hits ReportStream resource and we check the access token for authorization (now knowing that we can trust authentication)
- If the request is unauthorized, short-circuit the request to a 401
- If the request is authorized for the resource, continue to the business logic

```kotlin
// Microservice code

// handles all incoming requests
@RequestMapping("/**")
fun protectedReadEndpoint(
    @RequestBody(required = false) body: String,
    method: HttpMethod, 
    request: HttpServletRequest,
    response: HttpServletResponse
) {
    // this is all psudeocode!
    
    // check if we need to auth at all
    val authRequired = condigService.isAuthRequired(request.path, method)
    
    return if (authRequired) {
        // call out to authn service
        val authnResult = authService.checkToken(request)
        
        // possible results
        when {

>             authnResult == Authenticated && authzResult == Authorized -> httpClient.request(method, protectedEndpoint, body)

Author: take out block

- Cheap

Cons
- Custom authn/authz code is often vulnerable to attacks we may not have considered
- Additional work for us to write and maintain

## Protocol

We will be using OAuth 2.0 with OIDC. OAuth 2.0 is an industry-standard authorization protocol and OIDC is
another open standard on top of OAuth that adds an identity layer (authentication).

We will be using the **client credentials** flow which was designed with machine-to-machine communication in 
mind. See the image below for how the flow works generally.

![Client Credentials Flow](clientCredentialsFlow.png)

Users with emails and passwords will use the Interaction Code flow. Most of the interaction occurs by browser redirects between
the frontend and authorization server. From the platform perspective it will be handled similarly once the access token
is retrieved.

## Definitions

- **authn**: Authenticate/authentication
- **authz**: Authorize/authorization
- **IAM**: Identity and access management
- **IdP**: Identity Provider
- **Okta**: A cloud-hosted identity and access management service
- **Server2Server**: ReportStream's implementation JWT authn/authz
- **VPC**: Virtual Private Cloud

> - **Resource Server**: A server that will require authn/authz to access. In our case ResportStream

Reviewer: CTRL-F for `ResportStream` and fix spelling, occurs multiple times.

Author: I blame copy/paste! (or else I misspelled it "ResportStream" 5 times)

- **Authorization Server**: A server that handles authn/authz and issuing/revoking credentials
- **JWT Claim**: A key/value pair in the json body of the JWT
- **JWT Scopes**: A reserved claim in the [OAuth 2.0 spec](https://datatracker.ietf.org/doc/html/rfc8693#name-scope-scopes-claim)  containing a space-separated list of scopes associated with the token.

## Architecture options

### Spring reverse proxy microservice

This approach would create a new microservice whose sole responsibility is authn/authz. It would sit in front of
all requests to protected endpoints.

### Prerequisites
- Designate protected endpoints and put them behind a VPC.
  - Traffic must flow through our authn/authz service  
  - Protected endpoints should not be accessible to the public internet

> A typical report request would look like this

Reviewer 3: Should keep the sections in the same order for consistency.

    - In cloud-based solutions, we would make calls out to an external authz server
      - We could also introspect the token locally if speed is a factor but its more code to write/maintain
    - In a ReportStream implementation solution, we would introspect certain cookies and headers and compare them to what we have in a datastore
- If the request is unauthenticated or unauthorized, short-circuit the request to a 401
- If the request is authenticated and authorized for the resource, continue to the business logic

Code example:
```kotlin
// Spring project
// has read scope
@GetMapping("/api/v1/protected/endpoint")
@PreAuthorize("hasAuthority('SCOPE_read')") // scope 
fun protectedReadEndpoint() {
    // read that anyone with the read scope can see
}

// Spring project
// has admin authority
@PostMapping("/api/v1/protected/endpoint")
@PreAuthorize("hasAuthority('Admin')") // role
fun protectedWriteEndpoint() {
    // write that only admins can do
}

// Azure functions project
// for an Azure HTTP endpoint we could write a function to handle the authn/authz check
fun reportsEndpoint() {
    checkAuthn(request) { accessToken: JWT ->

>         requiredAuthzScope(accessToken, listOf("requiredScope")) {

Reviewer: Comment applies everywhere but putting it here:

We should remember that scope isn't the only criteria for determining authorization, in ReportStream. In the case of submitting reports, the client should have a `SCOPE` like "write" but ALSO their `CLAIMS` must include the `organization.sender` they are trying to submit as. Same applies for history endpoint: just because a client has the proper "read" SCOPE, doesn't mean they can get the history info for a report they didn't submit.

Reviewer: Okay never mind you discuss this in the section titled `### Scopes and claims` below but I will leave this comment here for documentation purposes. No action required.

// Spring project
// has read scope
@GetMapping("/api/v1/protected/endpoint")
@PreAuthorize("hasAuthority('SCOPE_read')") // scope 
fun protectedReadEndpoint() {
    // read that anyone with the read scope can see
}

// Azure functions project
// for an Azure HTTP endpoint we could write a function to handle the authn/authz check
fun reportsEndpoint() {
    val accessToken = request.headers("Authorization")
    requiredAuthzScope(accessToken, listOf("requiredScope")) {
        // business logic
    }
}
```

![Combo](combo.png)

###  Traditional Microservice

This approach would create a more traditional microservice that ReportStream would call out to when it required authn/authz functionality.

Pros
- Microservice able to handle parts of authentication that ReportStream will not have to worry about
- Authz code would be directly adjacent to the endpoint mappings they protect
- Can handle complex authorization requirements for specific endpoints with ease
- Less devops support as ReportStream would still handle incoming requests
- Easily Scalable to other services

Cons
- Authentication code must be called directly in ReportStream
- Unable to take advantage of annotations

> 

Reviewer 3: Missing the "A typical report request would look like this" section
